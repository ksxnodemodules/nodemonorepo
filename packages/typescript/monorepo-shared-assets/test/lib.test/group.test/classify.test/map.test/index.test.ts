import {iter, group} from '../../../../../index'
const classify = group.classify.map
const getList = () => iter.fns.range(16)
const factors = Array.from(iter.fns.range(2).up.to(16))

class ValueCarrier {
  readonly value: number

  constructor (value: number) {
    this.value = value
  }

  equal (cmp: ValueCarrier) {
    return this.value === cmp.value
  }
}

describe('single distribution', () => {
  it('in basic use', () => {
    const classifier = (x: number) => `Ring(4, ${x % 4})`
    expect(classify.singleDistribute(getList(), classifier)).toMatchSnapshot()
  })

  it('with specified identifier', () => {
    class RingOf4 {
      readonly shift: number

      constructor (shift: number) {
        this.shift = shift
      }

      equal (cmp: RingOf4) {
        return this.shift === cmp.shift
      }
    }

    const classifier = (x: number) => new RingOf4(x % 4)
    const identifier = (a: RingOf4, b: RingOf4) => a.equal(b)

    expect(classify.singleDistribute(getList(), classifier, identifier)).toMatchSnapshot()
  })

  it('with specified duplicationChecker', () => {
    const list = Array
      .from(iter.fns.repeat(getList(), 3))
      .map(value => new ValueCarrier(value))

    const classifier = (x: ValueCarrier) => `Ring(4, ${x.value % 4})`
    const duplicationChecker = (a: ValueCarrier, b: ValueCarrier) => a.equal(b)

    expect(classify.singleDistribute(list, classifier, undefined, duplicationChecker)).toMatchSnapshot()
  })
})

describe('multiple distribution', () => {
  it('in basic use', () => {
    const classifier = (x: number) => factors
      .filter(y => ![0, 1, x].includes(y) && !(x % y))
      .map(y => `x${y}`)

    expect(classify.multiDistribute(getList(), classifier)).toMatchSnapshot()
  })

  it('with specified identifier', () => {
    class Multiple {
      readonly factor: number

      constructor (factor: number) {
        this.factor = factor
      }

      equal (cmp: Multiple) {
        return this.factor === cmp.factor
      }
    }

    const classifier = (x: number) => factors
      .filter(factor => ![0, 1, x].includes(factor) && !(x % factor))
      .map(factor => new Multiple(factor))

    const identifier = (a: Multiple, b: Multiple) => a.equal(b)

    expect(classify.multiDistribute(getList(), classifier, identifier)).toMatchSnapshot()
  })

  it('with specified duplicationChecker', () => {
    const list = Array
      .from(iter.fns.repeat(getList(), 3))
      .map(value => new ValueCarrier(value))

    const classifier = (x: ValueCarrier) => factors
      .filter(y => ![0, 1, x.value].includes(y) && !(x.value % y))
      .map(y => `x${y}`)

    const duplicationChecker = (a: ValueCarrier, b: ValueCarrier) => a.equal(b)

    expect(classify.multiDistribute(list, classifier, undefined, duplicationChecker)).toMatchSnapshot()
  })
})
