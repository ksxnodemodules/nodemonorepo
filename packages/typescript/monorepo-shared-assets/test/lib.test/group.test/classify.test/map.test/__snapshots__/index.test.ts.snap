// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple distribution in basic use 1`] = `
Object {
  "classified": Map {
    "x2" => Set {
      0,
      4,
      6,
      8,
      10,
      12,
      14,
    },
    "x3" => Set {
      0,
      6,
      9,
      12,
      15,
    },
    "x4" => Set {
      0,
      8,
      12,
    },
    "x5" => Set {
      0,
      10,
      15,
    },
    "x6" => Set {
      0,
      12,
    },
    "x7" => Set {
      0,
      14,
    },
    "x8" => Set {
      0,
    },
    "x9" => Set {
      0,
    },
    "x10" => Set {
      0,
    },
    "x11" => Set {
      0,
    },
    "x12" => Set {
      0,
    },
    "x13" => Set {
      0,
    },
    "x14" => Set {
      0,
    },
    "x15" => Set {
      0,
    },
  },
  "unclassified": Set {
    1,
    2,
    3,
    5,
    7,
    11,
    13,
  },
}
`;

exports[`multiple distribution with specified duplicationChecker 1`] = `
Object {
  "classified": Map {
    "x2" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 4,
      },
      ValueCarrier {
        "value": 6,
      },
      ValueCarrier {
        "value": 8,
      },
      ValueCarrier {
        "value": 10,
      },
      ValueCarrier {
        "value": 12,
      },
      ValueCarrier {
        "value": 14,
      },
    },
    "x3" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 6,
      },
      ValueCarrier {
        "value": 9,
      },
      ValueCarrier {
        "value": 12,
      },
      ValueCarrier {
        "value": 15,
      },
    },
    "x4" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 8,
      },
      ValueCarrier {
        "value": 12,
      },
    },
    "x5" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 10,
      },
      ValueCarrier {
        "value": 15,
      },
    },
    "x6" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 12,
      },
    },
    "x7" => Set {
      ValueCarrier {
        "value": 0,
      },
      ValueCarrier {
        "value": 14,
      },
    },
    "x8" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x9" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x10" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x11" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x12" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x13" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x14" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
    "x15" => Set {
      ValueCarrier {
        "value": 0,
      },
    },
  },
  "unclassified": Set {
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 13,
    },
  },
}
`;

exports[`multiple distribution with specified identifier 1`] = `
Object {
  "classified": Map {
    Multiple {
      "factor": 2,
    } => Set {
      0,
      4,
      6,
      8,
      10,
      12,
      14,
    },
    Multiple {
      "factor": 3,
    } => Set {
      0,
      6,
      9,
      12,
      15,
    },
    Multiple {
      "factor": 4,
    } => Set {
      0,
      8,
      12,
    },
    Multiple {
      "factor": 5,
    } => Set {
      0,
      10,
      15,
    },
    Multiple {
      "factor": 6,
    } => Set {
      0,
      12,
    },
    Multiple {
      "factor": 7,
    } => Set {
      0,
      14,
    },
    Multiple {
      "factor": 8,
    } => Set {
      0,
    },
    Multiple {
      "factor": 9,
    } => Set {
      0,
    },
    Multiple {
      "factor": 10,
    } => Set {
      0,
    },
    Multiple {
      "factor": 11,
    } => Set {
      0,
    },
    Multiple {
      "factor": 12,
    } => Set {
      0,
    },
    Multiple {
      "factor": 13,
    } => Set {
      0,
    },
    Multiple {
      "factor": 14,
    } => Set {
      0,
    },
    Multiple {
      "factor": 15,
    } => Set {
      0,
    },
  },
  "unclassified": Set {
    1,
    2,
    3,
    5,
    7,
    11,
    13,
  },
}
`;

exports[`single distribution in basic use 1`] = `
Map {
  "Ring(4, 0)" => Set {
    0,
    4,
    8,
    12,
  },
  "Ring(4, 1)" => Set {
    1,
    5,
    9,
    13,
  },
  "Ring(4, 2)" => Set {
    2,
    6,
    10,
    14,
  },
  "Ring(4, 3)" => Set {
    3,
    7,
    11,
    15,
  },
}
`;

exports[`single distribution with specified duplicationChecker 1`] = `
Map {
  "Ring(4, 0)" => Set {
    ValueCarrier {
      "value": 0,
    },
    ValueCarrier {
      "value": 4,
    },
    ValueCarrier {
      "value": 8,
    },
    ValueCarrier {
      "value": 12,
    },
  },
  "Ring(4, 1)" => Set {
    ValueCarrier {
      "value": 1,
    },
    ValueCarrier {
      "value": 5,
    },
    ValueCarrier {
      "value": 9,
    },
    ValueCarrier {
      "value": 13,
    },
  },
  "Ring(4, 2)" => Set {
    ValueCarrier {
      "value": 2,
    },
    ValueCarrier {
      "value": 6,
    },
    ValueCarrier {
      "value": 10,
    },
    ValueCarrier {
      "value": 14,
    },
  },
  "Ring(4, 3)" => Set {
    ValueCarrier {
      "value": 3,
    },
    ValueCarrier {
      "value": 7,
    },
    ValueCarrier {
      "value": 11,
    },
    ValueCarrier {
      "value": 15,
    },
  },
}
`;

exports[`single distribution with specified identifier 1`] = `
Map {
  RingOf4 {
    "shift": 0,
  } => Set {
    0,
    4,
    8,
    12,
  },
  RingOf4 {
    "shift": 1,
  } => Set {
    1,
    5,
    9,
    13,
  },
  RingOf4 {
    "shift": 2,
  } => Set {
    2,
    6,
    10,
    14,
  },
  RingOf4 {
    "shift": 3,
  } => Set {
    3,
    7,
    11,
    15,
  },
}
`;
